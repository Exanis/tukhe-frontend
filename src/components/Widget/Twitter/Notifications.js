import React, { Component } from 'react';
import PropTypes from 'prop-types';

import List from '@material-ui/core/List';
import InfiniteScroll from 'react-infinite-scroller';
import Tweet from './Tweet';

export default class Notifications extends Component {
    state = {
        displayPages: 0
    };

    pseudoLoadData = (page) => this.setState({displayPages: page});

    render () {
        if (!this.props.data)
            return null;

        const items = this.props.data.slice(
            0, (this.state.displayPages + 1) * 10
        ).map(tweet => <Tweet
            updateElement={this.props.updateElement}
            actionWithAccount={this.props.actionWithAccount}
            displayElementInPopup={this.props.displayElementInPopup}
            tweet={tweet}
            key={tweet.id}
            updatePopupProps={this.props.updatePopupProps}
        />);

        return <List className={this.props.containerClass}>
            <InfiniteScroll
                pageStart={0}
                loadMore={this.pseudoLoadData}
                hasMore={(this.state.displayPages + 1) * 10 < this.props.data.length}
                useWindow={false}
                >
                {items}
            </InfiniteScroll>
        </List>
    }

    componentDidMount() {
        this.props.startWidgetProcess({
            target: this.props.refreshWithAccount,
            interval: 21000,
            widget: this.props.uuid
        });
    }

    componentWillUnmount() {
        this.props.stopWidgetProcess({
            widget: this.props.uuid
        })
    }
}

Notifications.propTypes = {
    data: PropTypes.array,
    updateElement: PropTypes.func.isRequired,
    actionWithAccount: PropTypes.func.isRequired,
    displayElementInPopup: PropTypes.func.isRequired,
    updatePopupProps: PropTypes.func.isRequired,
    containerClass: PropTypes.string,
    startWidgetProcess: PropTypes.func.isRequired,
    refreshWithAccount: PropTypes.func.isRequired,
    stopWidgetProcess: PropTypes.func.isRequired
};
