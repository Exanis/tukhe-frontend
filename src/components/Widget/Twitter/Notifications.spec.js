import React from 'react';
import { shallow } from 'enzyme';
import Notifications from './Notifications';

const props = {
    actionWithAccount: jest.fn(),
    addToWidgetData: jest.fn(),
    closeElementPopup: jest.fn(),
    updatePopupProps: jest.fn(),
    displayElementInPopup: jest.fn(),
    data: [{
        id: 'tweet number one',
        retweeted_status: null,
        text: "Test de \ntweet [badlink] [otherlink] [link] [media]",
        quoted: {
            id: 'tweetid',
        },
        urls: [
            {
                'target': 'twitter/tweetid',
                'short': '[link]'
            },
            {
                'target': 'http://www.google.fr',
                'short': '[badlink]'
            },
            {
                'target': 'http://www.twitter.com',
                'short': '[otherlink]'
            }
        ],
        media: [
            {
                url: '[media]',
                media_url: 'real media url'
            }
        ],
        author: {
            name: 'test',
            avatar: 'http://example.org/image.gif',
            screen_name: 'test screen name',
            protected: false,
            verified: true
        },
        user_liked: true,
        user_retweeted: true,
        replying_to: 'test reply',
    }],
    updateElement: jest.fn(),
    containerClass: 'test class',
    startWidgetProcess: jest.fn(),
    refreshWithAccount: jest.fn(),
    uuid: 'uuid',
    updateWidgetActions: jest.fn(),
    stopWidgetProcess: jest.fn(),
    intl: {
        formatMessage: () => 'test'
    }
};

it('Should render correctly', () => {
    const wrapper = shallow(<Notifications {...props} />);

    expect(wrapper).toMatchSnapshot();
    wrapper.unmount();
    expect(props.stopWidgetProcess).toHaveBeenCalledWith({
        widget: 'uuid'
    });
});

it('Should render correctly without data', () => {
    const testProps = {
        ...props,
        data: null
    };
    const wrapper = shallow(<Notifications {...testProps}/>);

    expect(wrapper).toMatchSnapshot();
});

it('Should correctly move between pages', () => {
    const wrapper = shallow(<Notifications {...props} />);

    wrapper.instance().pseudoLoadData(2);
    expect(wrapper.state('displayPages')).toEqual(2);
});
