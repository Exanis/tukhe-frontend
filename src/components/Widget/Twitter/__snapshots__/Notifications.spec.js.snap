// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Notifications
    actionWithAccount={[MockFunction]}
    addToWidgetData={[MockFunction]}
    closeElementPopup={[MockFunction]}
    containerClass="test class"
    data={
      Array [
        Object {
          "author": Object {
            "avatar": "http://example.org/image.gif",
            "name": "test",
            "protected": false,
            "screen_name": "test screen name",
            "verified": true,
          },
          "id": "tweet number one",
          "media": Array [
            Object {
              "media_url": "real media url",
              "url": "[media]",
            },
          ],
          "quoted": Object {
            "id": "tweetid",
          },
          "replying_to": "test reply",
          "retweeted_status": null,
          "text": "Test de 
tweet [badlink] [otherlink] [link] [media]",
          "urls": Array [
            Object {
              "short": "[link]",
              "target": "twitter/tweetid",
            },
            Object {
              "short": "[badlink]",
              "target": "http://www.google.fr",
            },
            Object {
              "short": "[otherlink]",
              "target": "http://www.twitter.com",
            },
          ],
          "user_liked": true,
          "user_retweeted": true,
        },
      ]
    }
    displayElementInPopup={[MockFunction]}
    intl={
      Object {
        "formatMessage": [Function],
      }
    }
    refreshWithAccount={[MockFunction]}
    startWidgetProcess={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "interval": 21000,
              "target": [MockFunction],
              "widget": "uuid",
            },
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    stopWidgetProcess={[MockFunction]}
    updateElement={[MockFunction]}
    updatePopupProps={[MockFunction]}
    updateWidgetActions={[MockFunction]}
    uuid="uuid"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <InfiniteScroll
        element="div"
        getScrollParent={null}
        hasMore={false}
        initialLoad={true}
        isReverse={false}
        loadMore={[Function]}
        loader={null}
        pageStart={0}
        ref={null}
        threshold={250}
        useCapture={false}
        useWindow={false}
      >
        <WithStyles(InjectIntl(RawTweet))
          actionWithAccount={[MockFunction]}
          displayElementInPopup={[MockFunction]}
          tweet={
            Object {
              "author": Object {
                "avatar": "http://example.org/image.gif",
                "name": "test",
                "protected": false,
                "screen_name": "test screen name",
                "verified": true,
              },
              "id": "tweet number one",
              "media": Array [
                Object {
                  "media_url": "real media url",
                  "url": "[media]",
                },
              ],
              "quoted": Object {
                "id": "tweetid",
              },
              "replying_to": "test reply",
              "retweeted_status": null,
              "text": "Test de 
tweet [badlink] [otherlink] [link] [media]",
              "urls": Array [
                Object {
                  "short": "[link]",
                  "target": "twitter/tweetid",
                },
                Object {
                  "short": "[badlink]",
                  "target": "http://www.google.fr",
                },
                Object {
                  "short": "[otherlink]",
                  "target": "http://www.twitter.com",
                },
              ],
              "user_liked": true,
              "user_retweeted": true,
            }
          }
          updateElement={[MockFunction]}
          updatePopupProps={[MockFunction]}
        />
      </InfiniteScroll>,
      "className": "test class",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <WithStyles(InjectIntl(RawTweet))
            actionWithAccount={[MockFunction]}
            displayElementInPopup={[MockFunction]}
            tweet={
              Object {
                "author": Object {
                  "avatar": "http://example.org/image.gif",
                  "name": "test",
                  "protected": false,
                  "screen_name": "test screen name",
                  "verified": true,
                },
                "id": "tweet number one",
                "media": Array [
                  Object {
                    "media_url": "real media url",
                    "url": "[media]",
                  },
                ],
                "quoted": Object {
                  "id": "tweetid",
                },
                "replying_to": "test reply",
                "retweeted_status": null,
                "text": "Test de 
tweet [badlink] [otherlink] [link] [media]",
                "urls": Array [
                  Object {
                    "short": "[link]",
                    "target": "twitter/tweetid",
                  },
                  Object {
                    "short": "[badlink]",
                    "target": "http://www.google.fr",
                  },
                  Object {
                    "short": "[otherlink]",
                    "target": "http://www.twitter.com",
                  },
                ],
                "user_liked": true,
                "user_retweeted": true,
              }
            }
            updateElement={[MockFunction]}
            updatePopupProps={[MockFunction]}
          />,
        ],
        "element": "div",
        "getScrollParent": null,
        "hasMore": false,
        "initialLoad": true,
        "isReverse": false,
        "loadMore": [Function],
        "loader": null,
        "pageStart": 0,
        "ref": null,
        "threshold": 250,
        "useCapture": false,
        "useWindow": false,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "tweet number one",
          "nodeType": "class",
          "props": Object {
            "actionWithAccount": [MockFunction],
            "displayElementInPopup": [MockFunction],
            "tweet": Object {
              "author": Object {
                "avatar": "http://example.org/image.gif",
                "name": "test",
                "protected": false,
                "screen_name": "test screen name",
                "verified": true,
              },
              "id": "tweet number one",
              "media": Array [
                Object {
                  "media_url": "real media url",
                  "url": "[media]",
                },
              ],
              "quoted": Object {
                "id": "tweetid",
              },
              "replying_to": "test reply",
              "retweeted_status": null,
              "text": "Test de 
tweet [badlink] [otherlink] [link] [media]",
              "urls": Array [
                Object {
                  "short": "[link]",
                  "target": "twitter/tweetid",
                },
                Object {
                  "short": "[badlink]",
                  "target": "http://www.google.fr",
                },
                Object {
                  "short": "[otherlink]",
                  "target": "http://www.twitter.com",
                },
              ],
              "user_liked": true,
              "user_retweeted": true,
            },
            "updateElement": [MockFunction],
            "updatePopupProps": [MockFunction],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <InfiniteScroll
          element="div"
          getScrollParent={null}
          hasMore={false}
          initialLoad={true}
          isReverse={false}
          loadMore={[Function]}
          loader={null}
          pageStart={0}
          ref={null}
          threshold={250}
          useCapture={false}
          useWindow={false}
        >
          <WithStyles(InjectIntl(RawTweet))
            actionWithAccount={[MockFunction]}
            displayElementInPopup={[MockFunction]}
            tweet={
              Object {
                "author": Object {
                  "avatar": "http://example.org/image.gif",
                  "name": "test",
                  "protected": false,
                  "screen_name": "test screen name",
                  "verified": true,
                },
                "id": "tweet number one",
                "media": Array [
                  Object {
                    "media_url": "real media url",
                    "url": "[media]",
                  },
                ],
                "quoted": Object {
                  "id": "tweetid",
                },
                "replying_to": "test reply",
                "retweeted_status": null,
                "text": "Test de 
tweet [badlink] [otherlink] [link] [media]",
                "urls": Array [
                  Object {
                    "short": "[link]",
                    "target": "twitter/tweetid",
                  },
                  Object {
                    "short": "[badlink]",
                    "target": "http://www.google.fr",
                  },
                  Object {
                    "short": "[otherlink]",
                    "target": "http://www.twitter.com",
                  },
                ],
                "user_liked": true,
                "user_retweeted": true,
              }
            }
            updateElement={[MockFunction]}
            updatePopupProps={[MockFunction]}
          />
        </InfiniteScroll>,
        "className": "test class",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <WithStyles(InjectIntl(RawTweet))
              actionWithAccount={[MockFunction]}
              displayElementInPopup={[MockFunction]}
              tweet={
                Object {
                  "author": Object {
                    "avatar": "http://example.org/image.gif",
                    "name": "test",
                    "protected": false,
                    "screen_name": "test screen name",
                    "verified": true,
                  },
                  "id": "tweet number one",
                  "media": Array [
                    Object {
                      "media_url": "real media url",
                      "url": "[media]",
                    },
                  ],
                  "quoted": Object {
                    "id": "tweetid",
                  },
                  "replying_to": "test reply",
                  "retweeted_status": null,
                  "text": "Test de 
tweet [badlink] [otherlink] [link] [media]",
                  "urls": Array [
                    Object {
                      "short": "[link]",
                      "target": "twitter/tweetid",
                    },
                    Object {
                      "short": "[badlink]",
                      "target": "http://www.google.fr",
                    },
                    Object {
                      "short": "[otherlink]",
                      "target": "http://www.twitter.com",
                    },
                  ],
                  "user_liked": true,
                  "user_retweeted": true,
                }
              }
              updateElement={[MockFunction]}
              updatePopupProps={[MockFunction]}
            />,
          ],
          "element": "div",
          "getScrollParent": null,
          "hasMore": false,
          "initialLoad": true,
          "isReverse": false,
          "loadMore": [Function],
          "loader": null,
          "pageStart": 0,
          "ref": null,
          "threshold": 250,
          "useCapture": false,
          "useWindow": false,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "tweet number one",
            "nodeType": "class",
            "props": Object {
              "actionWithAccount": [MockFunction],
              "displayElementInPopup": [MockFunction],
              "tweet": Object {
                "author": Object {
                  "avatar": "http://example.org/image.gif",
                  "name": "test",
                  "protected": false,
                  "screen_name": "test screen name",
                  "verified": true,
                },
                "id": "tweet number one",
                "media": Array [
                  Object {
                    "media_url": "real media url",
                    "url": "[media]",
                  },
                ],
                "quoted": Object {
                  "id": "tweetid",
                },
                "replying_to": "test reply",
                "retweeted_status": null,
                "text": "Test de 
tweet [badlink] [otherlink] [link] [media]",
                "urls": Array [
                  Object {
                    "short": "[link]",
                    "target": "twitter/tweetid",
                  },
                  Object {
                    "short": "[badlink]",
                    "target": "http://www.google.fr",
                  },
                  Object {
                    "short": "[otherlink]",
                    "target": "http://www.twitter.com",
                  },
                ],
                "user_liked": true,
                "user_retweeted": true,
              },
              "updateElement": [MockFunction],
              "updatePopupProps": [MockFunction],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`Should render correctly without data 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Notifications
    actionWithAccount={[MockFunction]}
    addToWidgetData={[MockFunction]}
    closeElementPopup={[MockFunction]}
    containerClass="test class"
    data={null}
    displayElementInPopup={[MockFunction]}
    intl={
      Object {
        "formatMessage": [Function],
      }
    }
    refreshWithAccount={[MockFunction]}
    startWidgetProcess={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "interval": 21000,
              "target": [MockFunction],
              "widget": "uuid",
            },
          ],
          Array [
            Object {
              "interval": 21000,
              "target": [MockFunction],
              "widget": "uuid",
            },
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    stopWidgetProcess={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "widget": "uuid",
            },
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    updateElement={[MockFunction]}
    updatePopupProps={[MockFunction]}
    updateWidgetActions={[MockFunction]}
    uuid="uuid"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): null,
  Symbol(enzyme.__nodes__): Array [
    null,
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
